File: LeetCode/68-text-justification/Solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {

  /**
   * Fully justifies a given list of words within a specified maxWidth.
   *
   * @param words Array of words to justify
   * @param maxWidth Maximum width of each justified line
   * @return List of fully justified lines
   */
  public List<String> fullJustify(String[] words, int maxWidth) {
    List<String> ans = new ArrayList<>();
    int i = 0;

    while (i < words.length) {
      // Gather words that fit within maxWidth for the current line
      List<String> currentLine = getWords(i, words, maxWidth);
      i += currentLine.size();

      // Format the selected words into a fully justified line
      ans.add(createLine(currentLine, i, words, maxWidth));
    }

    return ans;
  }

  /**
   * Selects words that fit within maxWidth for a single line.
   *
   * @param i Current word index
   * @param words Array of words to process
   * @param maxWidth Maximum width of each justified line
   * @return List of words forming the current line
   */
  private List<String> getWords(int i, String[] words, int maxWidth) {
    List<String> currentLine = new ArrayList<>();
    int currLength = 0;

    while (i < words.length && currLength + words[i].length() <= maxWidth) {
      currentLine.add(words[i]);
      // Account for space after each word except the last one
      currLength += words[i].length() + 1;
      i++;
    }

    return currentLine;
  }

  /**
   * Constructs a fully justified line from selected words.
   *
   * @param line Words forming the current line
   * @param i Index tracking processed words
   * @param words Array of all words
   * @param maxWidth Maximum width of the justified line
   * @return Fully justified line as a string
   */
  private String createLine(List<String> line, int i, String[] words, int maxWidth) {

    // Check if this is the last line or a single-word line
    boolean isLastLine = (i == words.length);
    boolean isSingleWord = (line.size() == 1);

    if (isLastLine || isSingleWord) {
      StringBuilder sb = new StringBuilder(String.join(" ", line));

      // Append spaces to reach maxWidth for left justification
      while (sb.length() < maxWidth) {
        sb.append(" ");
      }

      return sb.toString();
    }

    // Calculate base length excluding trailing space
    // Start at -1 to offset the initial space handling
    int baseLength = -1;
    for (String word : line) {
      baseLength += word.length() + 1; // Word length plus one space
    }

    // Compute the number of extra spaces to distribute
    int extraSpaces = maxWidth - baseLength;
    // Determine number of gaps between words
    int wordCount = line.size() - 1;
    // Compute evenly distributed spaces between words
    int spacesPerWord = extraSpaces / wordCount;
    // Assign remaining spaces to the leftmost words for balance
    int needsExtraSpaces = extraSpaces % wordCount;

    // Construct the fully justified line
    StringBuilder sb = new StringBuilder();

    for (int j = 0; j < line.size(); j++) {
      // Append the current word
      sb.append(line.get(j));

      // Add spaces between words, except after the last one
      if (j < wordCount) {
        for (int k = 0; k < spacesPerWord; k++) {
          sb.append(" ");
        }
        // Assign extra spaces to the leftmost words for even distribution
        if (j < needsExtraSpaces) {
          sb.append(" ");
        }
      }
    }

    return sb.toString();
  }
}

File: LeetCode/68-text-justification/solution.ts
/**
 * Fully justifies a given set of words within a specified maxWidth.
 * @param words - Array of words to justify.
 * @param maxWidth - Maximum width of each justified line.
 * @returns Array of fully justified lines.
 */
function fullJustify(words: string[], maxWidth: number): string[] {
  const result: string[] = [];
  let i = 0;

  while (i < words.length) {
    // Collect words that fit within maxWidth for the current line
    const currentLine = getWords(i, words, maxWidth);
    i += currentLine.length;

    // Format and justify the selected words
    result.push(createLine(currentLine, i, words, maxWidth));
  }

  return result;
}

/**
 * Selects words that fit within maxWidth for the current line.
 * @param i - Starting index in the words array.
 * @param words - Array of words to justify.
 * @param maxWidth - Maximum width of the line.
 * @returns Array of words forming a single justified line.
 */
function getWords(i: number, words: string[], maxWidth: number): string[] {
  const currentLine: string[] = [];
  let currLength = 0;

  while (i < words.length && currLength + words[i].length <= maxWidth) {
    currentLine.push(words[i]);
    // Include a space after each word
    currLength += words[i].length + 1;
    i++;
  }

  return currentLine;
}

/**
 * Formats a line of words into a justified string.
 * @param line - Array of words forming the current line.
 * @param i - Current index in the words array.
 * @param words - Array of words to justify.
 * @param maxWidth - Maximum width of the line.
 * @returns Fully justified line as a string.
 */
function createLine(
  line: string[],
  i: number,
  words: string[],
  maxWidth: number
): string {
  // Determine if this is the last line or a single-word line
  const isLastLine = i === words.length;
  const isSingleWord = line.length === 1;

  if (isLastLine || isSingleWord) {
    let justifiedString = line.join(" ");

    // Pad spaces to reach maxWidth for left justification
    while (justifiedString.length < maxWidth) {
      justifiedString += " ";
    }

    return justifiedString;
  }

  // Compute base length excluding trailing space
  // Start at -1 to offset initial space handling
  let baseLength = -1;
  for (const word of line) {
    baseLength += word.length + 1;
  }

  // Calculate the number of extra spaces to distribute
  const extraSpaces = maxWidth - baseLength;
  // Determine gaps between words
  const wordCount = line.length - 1;
  // Compute evenly distributed spaces between words
  const spacesPerWord = Math.floor(extraSpaces / wordCount);
  // Find remaining spaces for leftmost words
  const needsExtraSpaces = extraSpaces % wordCount;

  // Construct fully justified line
  let justifiedString = "";

  for (let j = 0; j < line.length; j++) {
    justifiedString += line[j];

    // Add spaces between words, except after the last one
    if (j < wordCount) {
      let totalSpaces = spacesPerWord + (j < needsExtraSpaces ? 1 : 0);
      
      justifiedString += (" ".repeat(totalSpaces));

      // Explicitly separate words to avoid merging issues
      if (j + 1 < line.length) {
        justifiedString += " ";
      }
    }
  }

  return justifiedString;
}